### PR TITLE
Optimistic-locking on diff

### PR DESCRIPTION
**Pull request by apelisse**
There is currently a race-condition when diffing, where we get the
object and then run a server-side dry-run patch and compare the two
results. If something changes the object on the server between the get
and the patch, the diff is going to show unrelated changes. We can now
specify the exact revisionversion that we want to patch, and that will
return a conflict, and we can retry multiple times to get a
non-conflicting diff. Eventually (after 3 times), we diff without
checking the version and throw a warning that the diff might be
partially wrong.

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/apply/apply.go [hunk 1]
```diff
@@ -824,7 +855,10 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 func (p *Patcher) Patch(current runtime.Object, modified []byte, source, namespace, name string, errOut io.Writer) ([]byte, runtime.Object, error) {
 	var getErr error
 	patchBytes, patchObject, err := p.patchSimple(current, modified, source, namespace, name, errOut)
-	for i := 1; i <= maxPatchRetry && errors.IsConflict(err); i++ {
+	if p.Retries == 0 {
+		p.Retries = 1
```
Comment (pwittrock) [inline pkg/kubectl/cmd/apply/apply.go]: shouldn't we keep this maxPatchRetry?

**Comment by AishSundar** [comment]
/cc @loburm @mikedanese 

@loburm we have a couple of PRs failing on pull-kubernetes-integration consistently today. Can you please help take a look to see if there's a wider GKE issue. this is critical PR blocked on getting into 1.13 release.

File: pkg/kubectl/cmd/apply/apply.go [hunk 2]
```diff
@@ -824,7 +855,10 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 func (p *Patcher) Patch(current runtime.Object, modified []byte, source, namespace, name string, errOut io.Writer) ([]byte, runtime.Object, error) {
 	var getErr error
 	patchBytes, patchObject, err := p.patchSimple(current, modified, source, namespace, name, errOut)
-	for i := 1; i <= maxPatchRetry && errors.IsConflict(err); i++ {
+	if p.Retries == 0 {
+		p.Retries = maxPatchRetry
+	}
+	for i := 1; i <= p.Retries && errors.IsConflict(err); i++ {
```
Comment (mikedanese) [inline pkg/kubectl/cmd/apply/apply.go]: use RetryOnConflict

https://github.com/kubernetes/kubernetes/blob/1431aeb45d4063a8125bda845d736dd748f49413/staging/src/k8s.io/client-go/util/retry/util.go#L61

File: pkg/kubectl/cmd/apply/apply.go [hunk 3]
```diff
@@ -812,6 +836,13 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return patch, obj, nil
 	}
 
+	if p.ResourceVersion != nil {
```
Comment (mikedanese) [inline pkg/kubectl/cmd/apply/apply.go]: Won't patchSimple will fail O(retries) times if `p.ResourceVersion != nil` and you get a conflict? Should you do this in Patch() before the first patchSimple and not enter the retry loop if `p.ResourceVersion != nil`?

File: pkg/kubectl/cmd/diff/diff.go [hunk 4]
```diff
@@ -376,21 +397,36 @@ func RunDiff(f cmdutil.Factory, diff *DiffProgram, options *DiffOptions) error {
 		}
 
 		local := info.Object.DeepCopyObject()
-		if err := info.Get(); err != nil {
-			if !errors.IsNotFound(err) {
-				return err
+		for i := 1; i <= maxRetries; i++ {
```
Comment (mikedanese) [inline pkg/kubectl/cmd/diff/diff.go]: Use RetryOnConflict? force should work fine.

**Pull request merged**